File: Leetcode/inorder_traversal.py
# https://leetcode.com/problems/binary-tree-inorder-traversal/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    13/10/2019
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        result = list()
        self.inorderTraversalRecursive(root, result)
        return result

    def inorderTraversalIterative(self, root: TreeNode, result: list) -> None:
        if(root == None):
            return
        stack = [root]

        while(len(stack) > 0):
            node = stack[-1]
            if(node.right == None and node.left == None):
                result.append(stack.pop().val)
                continue

            if(node.right != None):
                stack.append(node.right)
            if(node.left != None):
                stack.append(node.left)

    def inorderTraversalRecursive(self, root: TreeNode, result: list) -> None:
        if(root == None):
            return

        if(root.left != None):
            self.inorderTraversalRecursive(root.left, result)

        result.append(root.val)

        if(root.right != None):
            self.inorderTraversalRecursive(root.right, result)

File: InterviewBit/InorderTraversal.java
// https://www.interviewbit.com/problems/inorder-traversal/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    26/01/2021
*/

/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) {
 *      val = x;
 *      left=null;
 *      right=null;
 *     }
 * }
 */

import java.util.ArrayList;
import java.util.Stack;

public class Solution {
    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<>();

        TreeNode current = root;
        while (current != null) {
            stack.push(current);
            current = current.left;
        }


        while (!stack.empty()){
            current = stack.pop();
            
            result.add(current.val);
            current = current.right;
            while (current != null) {
                stack.push(current);
                current = current.left;
            }
        }

        return result;
    }
}
